source/common/init/watcher_impl.cc

```diff
@@ -4,34 +4,38 @@ namespace Envoy {
 namespace Init {
 
 WatcherHandleImpl::WatcherHandleImpl(absl::string_view handle_name, absl::string_view name,
-                                     std::weak_ptr<ReadyFn> fn)
+                                     std::weak_ptr<TargetAwareReadyFn> fn)
     : handle_name_(handle_name), name_(name), fn_(std::move(fn)) {}
 
 bool WatcherHandleImpl::ready() const {
   auto locked_fn(fn_.lock());
   if (locked_fn) {
     // If we can "lock" a shared pointer to the watcher's callback function, call it.
     ENVOY_LOG(debug, "{} initialized, notifying {}", handle_name_, name_);
-    (*locked_fn)();
+    (*locked_fn)(handle_name_);
     return true;
   } else {
     // If not, the watcher was already destroyed.
-    ENVOY_LOG(debug, "{} initialized, but can't notify {} (unavailable)", handle_name_, name_);
+    ENVOY_LOG(debug, "{} initialized, but can't notify {}", handle_name_, name_);
     return false;
   }
 }
 
 WatcherImpl::WatcherImpl(absl::string_view name, ReadyFn fn)
-    : name_(name), fn_(std::make_shared<ReadyFn>(std::move(fn))) {}
+    : name_(name), fn_(std::make_shared<TargetAwareReadyFn>(
+                       [callback = std::move(fn)](absl::string_view) { callback(); })) {}
+
+WatcherImpl::WatcherImpl(absl::string_view name, TargetAwareReadyFn fn)
+    : name_(name), fn_(std::make_shared<TargetAwareReadyFn>(std::move(fn))) {}
```

Comment:
Anyway, I removed the separate class ```TargetAwareWatcher```. Now the ```WatcherImpl``` can take both types of function in its Ctor. And the old one ```ReadyFn``` will be wrapped into the new type ```TargetAwareReadyFn```. Really appreciate your patience reviewing my code~.